File: rust/src/bin/sasatest.rs
use std::fs;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};

use nalgebra::Vector3;
use powersasa_rust::{PowerSasa, PowerSasaSettings};

#[derive(Debug, Clone)]
struct GoldenCase {
    file_path: String,
    expected_sasa: f32,
    expected_vol: f32,
}

fn load_case_file(path: &str) -> Result<(Vec<Vector3<f32>>, Vec<f32>), String> {
    let file = fs::File::open(path).map_err(|e| format!("Could not open {path}: {e}"))?;
    let reader = BufReader::new(file);

    let mut coords = Vec::new();
    let mut weights = Vec::new();

    for line in reader.lines() {
        let line = line.map_err(|e| format!("Error reading {path}: {e}"))?;
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let mut parts = line.split_whitespace();
        let Some(x) = parts.next() else { continue };
        let Some(y) = parts.next() else { continue };
        let Some(z) = parts.next() else { continue };
        let Some(radius) = parts.next() else { continue };

        let x: f32 = x.parse().map_err(|e| format!("Bad x in {path}: {e}"))?;
        let y: f32 = y.parse().map_err(|e| format!("Bad y in {path}: {e}"))?;
        let z: f32 = z.parse().map_err(|e| format!("Bad z in {path}: {e}"))?;
        let radius: f32 = radius
            .parse()
            .map_err(|e| format!("Bad radius in {path}: {e}"))?;

        coords.push(Vector3::new(x, y, z));
        weights.push(radius + 0.14);
    }

    if coords.is_empty() {
        return Err(format!("No atoms loaded from {path}"));
    }

    Ok((coords, weights))
}

fn load_golden_manifest(manifest_path: &Path) -> Result<Vec<GoldenCase>, String> {
    let file = fs::File::open(manifest_path)
        .map_err(|e| format!("Could not open manifest {}: {e}", manifest_path.display()))?;
    let reader = BufReader::new(file);

    let manifest_dir = manifest_path
        .parent()
        .ok_or_else(|| format!("Manifest has no parent: {}", manifest_path.display()))?;
    let repo_root = manifest_dir
        .parent()
        .and_then(|p| p.parent())
        .ok_or_else(|| format!("Could not derive repo root from {}", manifest_path.display()))?;

    let mut cases = Vec::new();

    for line in reader.lines() {
        let line = line.map_err(|e| format!("Error reading manifest {}: {e}", manifest_path.display()))?;
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let mut parts = line.split_whitespace();
        let Some(file_path) = parts.next() else { continue };
        let Some(expected_sasa) = parts.next() else { continue };
        let Some(expected_vol) = parts.next() else { continue };

        let expected_sasa: f32 = expected_sasa
            .parse()
            .map_err(|e| format!("Bad expected_sasa '{expected_sasa}': {e}"))?;
        let expected_vol: f32 = expected_vol
            .parse()
            .map_err(|e| format!("Bad expected_vol '{expected_vol}': {e}"))?;

        let listed = PathBuf::from(file_path);
        let resolved = if listed.is_absolute() {
            listed
        } else if listed.exists() {
            listed
        } else if (repo_root.join(&listed)).exists() {
            repo_root.join(&listed)
        } else {
            manifest_dir.join(&listed)
        };

        cases.push(GoldenCase {
            file_path: resolved.to_string_lossy().into_owned(),
            expected_sasa,
            expected_vol,
        });
    }

    if cases.is_empty() {
        return Err(format!("No cases in manifest {}", manifest_path.display()));
    }

    Ok(cases)
}

fn find_manifest() -> PathBuf {
    let rel = PathBuf::from("testdata/sasa_cases/golden_912340c.txt");
    for candidate in [rel.clone(), PathBuf::from("../").join(&rel), PathBuf::from("../../").join(&rel)] {
        if candidate.exists() {
            return candidate;
        }
    }
    rel
}

fn main() -> Result<(), String> {
    const EPSILON: f32 = 1e-3;
    let manifest = find_manifest();
    println!("Precision: float");
    println!("Manifest: {}", manifest.display());

    let cases = load_golden_manifest(&manifest)?;
    let mut all_passed = true;

    for c in &cases {
        let (coords, weights) = load_case_file(&c.file_path)?;

        let mut ps = PowerSasa::<f32>::new(
            coords.clone().into_iter(),
            weights.clone().into_iter(),
            PowerSasaSettings {
                with_sasa: true,
                with_dsasa: false,
                with_vol: true,
                with_dvol: false,
            },
        );
        ps.calc_sasa_all().map_err(|e| format!("calc_sasa_all failed: {e}"))?;

        let total_sasa: f32 = ps.get_sasa().iter().copied().sum();
        let total_vol: f32 = ps.get_vol().iter().copied().sum();

        let sasa_ok = (total_sasa - c.expected_sasa).abs() <= EPSILON;
        let vol_ok = (total_vol - c.expected_vol).abs() <= EPSILON;
        all_passed = all_passed && sasa_ok && vol_ok;

        println!("Case: {}", c.file_path);
        println!(
            "  SASA:   {:>12.6}  expected {:>12.6}  [{}]",
            total_sasa,
            c.expected_sasa,
            if sasa_ok { "OK" } else { "FAIL" }
        );
        println!(
            "  Volume: {:>12.6}  expected {:>12.6}  [{}]",
            total_vol,
            c.expected_vol,
            if vol_ok { "OK" } else { "FAIL" }
        );
    }

    if !all_passed {
        return Err("REGRESSION FAILURE: one or more cases differ from 912340c golden values.".to_string());
    }

    println!("REGRESSION PASS: all cases match 912340c golden values.");
    Ok(())
}

File: legacy/current/sasatest.cpp
#include <cmath>
#include <filesystem>
#include <fstream>
#include <numeric>
#include <print>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

#include <power_sasa.h>
#include "vec3.h"

namespace {
namespace fs = std::filesystem;

template <class Scalar, class Coord>
bool load_case_file(const std::string& path, std::vector<Coord>& coords, std::vector<Scalar>& weights)
{
	coords.clear();
	weights.clear();

	std::ifstream infile(path);
	if (!infile.is_open()) {
		std::cerr << "Error: Could not open " << path << '\n';
		return false;
	}

	std::string line;
	while (std::getline(infile, line)) {
		if (line.empty() || line[0] == '#') {
			continue;
		}
		std::istringstream iss(line);
		Scalar x, y, z, radius;
		if (!(iss >> x >> y >> z >> radius)) {
			continue;
		}
		coords.push_back(Coord(x, y, z));
		weights.push_back(radius + static_cast<Scalar>(0.14));
	}

	return !coords.empty();
}

struct GoldenCase {
	std::string filePath;
	float expectedSasa = 0.0f;
	float expectedVol = 0.0f;
};

bool load_golden_manifest(const fs::path& manifestPath, std::vector<GoldenCase>& cases)
{
	cases.clear();
	std::ifstream infile(manifestPath);
	if (!infile.is_open()) {
		std::cerr << "Error: Could not open manifest " << manifestPath << '\n';
		return false;
	}

	const fs::path manifestDir = manifestPath.parent_path();
	const fs::path repoRoot = manifestDir.parent_path().parent_path();
	std::string line;
	while (std::getline(infile, line)) {
		if (line.empty() || line[0] == '#') {
			continue;
		}
		std::istringstream iss(line);
		GoldenCase c;
		if (!(iss >> c.filePath >> c.expectedSasa >> c.expectedVol)) {
			std::cerr << "Error: Malformed manifest entry: " << line << '\n';
			return false;
		}
		const fs::path listed = c.filePath;
		if (!listed.is_absolute()) {
			if (fs::exists(listed)) {
				c.filePath = listed.lexically_normal().string();
			} else if (fs::exists(repoRoot / listed)) {
				c.filePath = (repoRoot / listed).lexically_normal().string();
			} else if (fs::exists(manifestDir / listed)) {
				c.filePath = (manifestDir / listed).lexically_normal().string();
			} else {
				c.filePath = (manifestDir / listed).lexically_normal().string();
			}
		}
		cases.push_back(c);
	}

	return !cases.empty();
}

fs::path find_manifest()
{
	constexpr std::string_view kRel = "tests/testdata/sasa_cases/golden_912340c.txt";
	const fs::path candidates[] = {
	    fs::path(kRel),
	    fs::path("..") / kRel,
	    fs::path("../..") / kRel,
	    fs::path("../../..") / kRel,
	};
	for (const fs::path& candidate : candidates) {
		if (fs::exists(candidate)) {
			return candidate;
		}
	}
	return fs::path(kRel);
}

}  // namespace

int main()
{
	using Scalar = float;
	using Coord = Vec3<Scalar>;
	constexpr Scalar epsilon = 1e-3f;
	const fs::path goldenManifest = find_manifest();
	std::println("Precision: {}", std::string_view{"float"});
	std::println("Manifest: {}", goldenManifest.string());

	std::vector<GoldenCase> cases;
	if (!load_golden_manifest(goldenManifest, cases)) {
		return 1;
	}

	bool allPassed = true;
	std::vector<Coord> coords;
	std::vector<Scalar> weights;

	for (const GoldenCase& c : cases) {
		if (!load_case_file<Scalar, Coord>(c.filePath, coords, weights)) {
			std::cerr << "Error: Failed to load atoms from " << c.filePath << '\n';
			return 1;
		}

		POWERSASA::PowerSasa<Scalar, Coord> ps(coords, weights, 1, 0, 1, 0);
		ps.calc_sasa_all();

		const std::vector<Scalar>& sasa = ps.getSasa();
		const std::vector<Scalar>& vol = ps.getVol();

		const Scalar totalSasa = std::accumulate(sasa.begin(), sasa.end(), Scalar(0));
		const Scalar totalVol = std::accumulate(vol.begin(), vol.end(), Scalar(0));

		const bool sasaOk = std::abs(totalSasa - c.expectedSasa) <= epsilon;
		const bool volOk = std::abs(totalVol - c.expectedVol) <= epsilon;
		allPassed = allPassed && sasaOk && volOk;

		std::println("Case: {}", c.filePath);
		std::println("  Atoms: {}", coords.size());
		std::println("  SASA:   {:>12.6f}  expected {:>12.6f}  [{}]",
		             totalSasa, c.expectedSasa, sasaOk ? "OK" : "FAIL");
		std::println("  Volume: {:>12.6f}  expected {:>12.6f}  [{}]",
		             totalVol, c.expectedVol, volOk ? "OK" : "FAIL");
	}

	if (!allPassed) {
		std::cerr << "REGRESSION FAILURE: one or more cases differ from 912340c golden values." << '\n';
		return 1;
	}

	std::println("REGRESSION PASS: all cases match 912340c golden values.");
	return 0;
}
